beps/bep_0003.rst

```diff
@@ -87,70 +92,153 @@ announce
 
 info
   This maps to a dictionary, with keys described below.
+  
+``piece layers``
+  A dictionary of strings. For each file in the file tree that is larger than the piece size
+  it contains one string value.
+  The keys are the merkle roots while the values consist of concatenated hashes
+  of one layer within that merkle tree.
+  The layer is chosen so that one hash covers ``piece length`` bytes.
+  For example if the piece size is 16KiB then the leaf hashes are used.
+  If a piece size of 128KiB is used then 3rd layer up from the leaf hashes is used.
+  Layer hashes which exclusively cover data beyond the end of file,
+  i.e. are only needed to balance the tree, are omitted.
+  All hashes are stored in their binary format.
+  
+  A torrent is not valid if this field is absent, the contained hashes do not match
+  the merkle roots or are not from the correct layer.
```

Comment:
That requires changes to the wire protocol sans extensions. If we allow omitting the layers then clients may be unable to download a torrent unless there is another way to transfer the hashes. My goal was to leave it unchanged if we take clients with fast extensions as baseline.

In other words it would increase the complexity of the core protocol to enable a use-case which currently is not covered by core-bittorrent but only by bittorrent-with-extensions while possibly impairing usecases that currently are guaranteed to work with core-bittorrent.

I really don't like that.